### PR TITLE
Add inter-foil distance feature to spatial bias routing

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The spatial_bias MLP currently routes slice attention based on (x, y, curvature, dist-to-surface, gap, stagger) — 6 scalars. These describe the **node's local geometry** and the **global tandem configuration**, but not the node's relationship to the *other* foil. We hypothesize that adding `log(1 + d_interfoil)` — the log-distance from each mesh node to foil-2's geometric center — as a 7th input to the spatial_bias MLP will allow the model to partition attention slices by proximity to foil-2, improving wake-interference predictions on tandem configurations.

Physical motivation: in tandem foils, pressure perturbations decay as ~1/r from the upstream foil. A node's log-distance to foil-2 is a meaningful geometric signal for distinguishing near-wake vs far-field regions, which the current 6-dim routing input cannot express. This is a purely structural enhancement — no feature distribution manipulation, no domain alignment.

The gap scalar (x[:, 0, 22]) and stagger scalar (x[:, 0, 23]) together define foil-2's approximate center offset relative to foil-1. The inter-foil distance for each node is then the Euclidean distance from (node_x, node_y) to (stagger, gap).

For **single-foil samples** (where gap≈0, stagger≈0), set d_interfoil = 10.0 (a large constant), so log1p(10) ≈ 2.4 is a stable sentinel that distinguishes single-foil from tandem samples without disrupting gradient flow.

## Instructions

Make the following changes to `cfd_tandemfoil/train.py`:

### Step 1: Add `--interfoil_dist_feature` flag

In the argparse section, add:
```python
parser.add_argument("--interfoil_dist_feature", action="store_true",
                    help="Add log(1+d_interfoil) as 7th spatial_bias input channel")
```

### Step 2: Pass flag through to `TransolverModel` / `Physics_Attention_Irregular_Mesh`

Thread `interfoil_dist_feature=args.interfoil_dist_feature` through the model constructor call and store as `self.interfoil_dist_feature`.

### Step 3: Extend spatial_bias input dim from 6→7

In the block/layer init, wherever `spatial_bias_input_dim` is set:
```python
spatial_bias_input_dim = 6 if gap_stagger_spatial_bias else 4
if interfoil_dist_feature and gap_stagger_spatial_bias:
    spatial_bias_input_dim = 7
```

### Step 4: Zero-init the new 7th input column

After the existing zero-init block for the `gap_stagger_spatial_bias` extra columns, add:
```python
if interfoil_dist_feature and gap_stagger_spatial_bias:
    with torch.no_grad():
        for block in self.blocks:
            block.spatial_bias[0].weight[:, 6:].zero_()
```
This ensures that at epoch 0, the new feature contributes zero, preserving the trained routing from the baseline.

### Step 5: Compute and append d_interfoil in the forward pass

In the forward pass, right after `raw_xy` is constructed (currently lines ~887-892), append the inter-foil distance:

```python
if self.interfoil_dist_feature and self.gap_stagger_spatial_bias:
    # gap_stagger: [B, N, 2] already computed above as x[:,0:1,22:24].expand(...)
    # gap = stagger scalar (index 22), stagger = stagger scalar (index 23)
    foil2_center = x[:, 0:1, 22:24]  # [B, 1, 2]: (gap, stagger) = approx foil-2 offset
    node_xy = x[:, :, :2]  # [B, N, 2]
    # For single-foil samples, gap≈0 and stagger≈0, so distance ≈ ||(node_x, node_y)||
    # Use is_tandem mask to replace with sentinel 10.0 for single-foil
    is_tandem_1d = (x[:, 0, 21].abs() > 0.01).float()  # [B]
    d_sq = ((node_xy - foil2_center) ** 2).sum(dim=-1, keepdim=True)  # [B, N, 1]
    d_interfoil = torch.sqrt(d_sq + 1e-8)  # [B, N, 1]
    # Replace single-foil distances with sentinel 10.0
    sentinel = torch.full_like(d_interfoil, 10.0)
    mask = is_tandem_1d[:, None, None]  # [B, 1, 1]
    d_interfoil = d_interfoil * mask + sentinel * (1.0 - mask)
    log_d = torch.log1p(d_interfoil)  # [B, N, 1]
    raw_xy = torch.cat([raw_xy, log_d], dim=-1)  # [B, N, 7]
```

### Step 6: Run with `--interfoil_dist_feature` alongside all baseline flags

Use `--wandb_group round7/interfoil-dist-feature` and run 2 seeds (42 and 73):

**Seed 42:**
```bash
cd cfd_tandemfoil && python train.py \
  --agent askeladd --wandb_name "askeladd/interfoil-dist-seed42" \
  --wandb_group round7/interfoil-dist-feature \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --interfoil_dist_feature
```

**Seed 73:**
```bash
cd cfd_tandemfoil && python train.py \
  --agent askeladd --wandb_name "askeladd/interfoil-dist-seed73" \
  --wandb_group round7/interfoil-dist-feature \
  --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --interfoil_dist_feature
```

### Step 7: Report results

When both seeds are done, add a **Results** comment to this PR with:
- 2-seed average of: p_in, p_oodc, p_tan, p_re
- Individual per-seed values
- W&B run IDs and links
- Whether any seeds diverged

## Baseline

Current best (PR #2184, DCT Freq Loss, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in | 13.205 | < 13.21 |
| p_oodc | 7.816 | < 7.82 |
| **p_tan** | **28.502** | **< 28.50** |
| p_re | 6.453 | < 6.45 |

W&B runs: 6yfv5lio (seed 42, p_tan=28.432), etepxvjc (seed 73, p_tan=28.572)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-dct-freq" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Summary

The inter-foil distance feature produces **mixed results**: improves p_in (-2.3%) and p_re (-1.6%), but the primary target metric p_tan slightly regresses (+2.4%). p_oodc is essentially flat.

### Individual Run Metrics (Surface MAE p, denormalized)

| Seed | W&B Run | p_in | p_tan | p_oodc | p_re | Best Epoch |
|------|---------|------|-------|--------|------|------------|
| 42 | `gf94dd2t` | 12.9 | 28.7 | 7.7 | 6.3 | 148 |
| 73 | `x5l2mf4g` | 12.9 | 29.7 | 7.9 | 6.4 | 148 |

### 2-Seed Averages vs Baseline

| Config | p_in | p_tan | p_oodc | p_re |
|--------|------|-------|--------|------|
| **Baseline (DCT Freq)** | **13.205** | **28.502** | **7.816** | **6.453** |
| + interfoil_dist_feature | 12.90 (-2.3%) | 29.20 (+2.4%) | 7.80 (-0.2%) | 6.35 (-1.6%) |

### Train Commands

```bash
cd cfd_tandemfoil && python train.py \
  --agent askeladd --wandb_name "askeladd/interfoil-dist-seed42" \
  --wandb_group round7/interfoil-dist-feature --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --interfoil_dist_feature
```

### Peak Memory

~43 GB VRAM per run (identical to baseline — negligible overhead from 1 extra input channel)

### W&B Run IDs

- `gf94dd2t` (seed 42), `x5l2mf4g` (seed 73)
- W&B group: `round7/interfoil-dist-feature`

### What happened

**Mixed results — improves in-distribution and OOD-Re but slightly hurts tandem transfer.**

1. **p_in improved by 2.3%** (12.90 vs 13.205) — the inter-foil distance provides useful geometric signal for the spatial bias MLP, helping it better partition attention slices for in-distribution tandem samples.

2. **p_re improved by 1.6%** (6.35 vs 6.453) — same benefit for OOD Reynolds number generalization.

3. **p_tan regressed by 2.4%** (29.20 vs 28.502) — this is the key target metric and it got worse. Seed 42 (p_tan=28.7) was close to baseline seed 42 (28.432), but seed 73 (p_tan=29.7) diverged significantly from baseline seed 73 (28.572). The regression is mostly driven by seed 73.

4. **Why p_tan regressed:** The inter-foil distance is computed using gap/stagger to estimate foil-2's center position. For OOD NACA6416 tandem configurations, the relationship between (gap, stagger) and the actual aerodynamic interaction zone may differ from training distributions. The model may be overfitting its slice routing to a tandem-specific distance pattern that doesn't generalize to the OOD foil shape.

5. **Implementation was correct:** Zero-init on the 7th column ensured baseline-equivalent behavior at epoch 0, and the feature gradually learned its contribution. The `log1p` transform and sentinel value (10.0 for single-foil) worked as expected.

### Note on verification NaN bug

Same pre-existing verification NaN issue as in prior PRs.

### Suggested follow-ups

1. **Normalize d_interfoil by chord length**: The raw distance depends on absolute scale. Normalizing by the local chord length would make it scale-invariant, potentially improving OOD generalization to different foil geometries.

2. **Use d_interfoil relative to foil-1 surface** instead of foil-2 center: The distance to the nearest point on foil-1's surface (from DSDF) may be more physically meaningful than distance to foil-2's center for wake interaction modeling.

3. **Combine with other improvements:** The p_in and p_re gains suggest this feature does capture useful geometry. Consider combining with techniques that specifically improve p_tan (e.g., those targeting the decoder/output side) to get compound benefit.

4. **The seed 73 outlier** (p_tan=29.7) warrants investigation. Check whether the spatial bias MLP learned different routing patterns across seeds — the feature may introduce optimization instability.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2195 because ADVISOR: **Closing — p_tan regressed +2.4% (29.20 vs 28.502 baseline).**

Good experiment, clean implementation. The zero-init strategy worked correctly, and the p_in (-2.3%) and p_re (-1.6%) improvements confirm the feature captures useful geometry. But p_tan — our primary target — got worse, especially seed 73 (p_tan=29.7 vs baseline 28.572).

**Root cause:** The inter-foil distance is computed from (gap, stagger) → foil-2 center. For OOD NACA6416, the relationship between this center offset and actual wake interaction zones differs from NACA0012 training distribution. The spatial bias MLP overfits its slice routing to training-specific distance patterns that don't transfer.

**Takeaway:** Adding per-node spatial features to the spatial_bias MLP can improve in-distribution performance but tends to hurt OOD tandem transfer. This direction (raw geometric distance features in spatial bias) is a dead end for p_tan. The gap/stagger scalars already capture sufficient tandem configuration information.

Your suggestion about normalizing by chord length is interesting but wouldn't fix the fundamental issue — the distance-to-center proxy doesn't capture the actual aerodynamic coupling zone shape, which changes with foil geometry.

I'll assign you a new experiment targeting the aft-foil wake coupling mechanism directly. Thanks for the thorough analysis..
